snuba/query/snql/parser.py

```diff
@@ -108,38 +141,78 @@ class OrTuple(NamedTuple):
     exp: Expression
 
 
+class EntityTuple(NamedTuple):
+    alias: str
+    name: str
+
+
 class SnQLVisitor(NodeVisitor):
     """
     Builds Snuba AST expressions from the Parsimonious parse tree.
     """
 
     def visit_query_exp(self, node: Node, visited_children: Iterable[Any]) -> Query:
-        _, where, collect, groupby, having, orderby = visited_children
+        match, where, collect, groupby, having, orderby, limit = visited_children
         # check for empty clauses
         if isinstance(groupby, Node):
             groupby = None
         if isinstance(orderby, Node):
             orderby = None
         if isinstance(having, Node):
             having = None
+        if isinstance(limit, Node):
+            limit = None
+
+        selected_columns = collect
+        _groupby = None
+        if groupby:
+            selected_columns += groupby
+            _groupby = [g.expression for g in groupby]
+
         return Query(
             body={},
-            from_clause=None,
-            selected_columns=collect,
+            from_clause=match,
+            selected_columns=selected_columns,
             array_join=None,
             condition=where,
             prewhere=None,
-            groupby=groupby,
+            groupby=_groupby,
             having=having,
             order_by=orderby,
+            limit=limit,
         )
 
+    def visit_match_clause(
+        self, node: Node, visited_children: Tuple[Any, Any, Any, EntityTuple, Any],
+    ) -> QueryEntity:
+        _, _, _, match, _ = visited_children
+        key = EntityKey(match.name)
+        query_entity = QueryEntity(key, get_entity(key).get_data_model())
+        return query_entity
+
+    def visit_entity_match(
+        self, node: Node, visited_children: Tuple[Any, str, Any, Any, str, Any],
+    ) -> EntityTuple:
+        _, alias, _, _, name, _ = visited_children
+        return EntityTuple(alias, name.lower())
+
+    def visit_entity_alias(self, node: Node, visited_children: Tuple[Any]) -> str:
+        return str(node.text)
+
+    def visit_entity_name(self, node: Node, visited_children: Tuple[Any]) -> str:
+        return str(node.text)
+
     def visit_function_name(self, node: Node, visited_children: Iterable[Any]) -> str:
         return visit_function_name(node, visited_children)
 
     def visit_column_name(self, node: Node, visited_children: Iterable[Any]) -> Column:
         return visit_column_name(node, visited_children)
 
+    def visit_subscriptable(
+        self, node: Node, visited_children: Iterable[Any]
+    ) -> Column:
+        return visit_column_name(node, visited_children)
```

Comment:
continuation from the comment above on the token to use for subscriptables.
I would just go ahead and create the `SubscriptableReference` here:
1) it is a valid node type in the AST, we are not creating a complex expression with a lot of logic involved, so it feels like a primitive object that can be built by the parser. Ok this can be subjective
2) more important. I think we should keep as many language concerns as possible inside the parser.  The syntax `tags[key]` could arguably considered language syntactic sugar. So the earlier we turn that into the appropriate AST the better. You also save one AST pass to do the transformation
3) I had to do the transformation in its own step (at least as far as I remember but I don't think there was any other reason) to ensure a subscriptable reference would have the appropriate alias during alias expansion (there was no way to rebuild tags[key] from the subscriptable reference without the AST knowing about the language). Now that is not an issue anymore, aliases do not drive the name of the column in the result anymore, so you should be able to parse it here and assign whatever alias you want to the expression.

Reply:
I had planned to make this change already. It came up when we were talking about the extra query function `_parse_subscriptables`. I didn't do it here just because there was going to be another PR where I rewrote these query parsing functions, and I was planning to move the logic from `_parse_subscriptables` into the grammar itself.